### DISCUSSION TITLE
Using Matrix Variable In Docker Image Name

### DISCUSSION THREAD

**Discussion by lots0logs** [community/community#25079] [Actions]
<p>I’ve seen a few other posts about this but I really don’t think any of the responses are good enough (unless I am missing something). How does GitHub recommend running tests using multiple languages and/or language versions when you have a different docker image for each one? For example, I need to run tests under PHP 5.2, 5.6, and 7.4. I have docker images hosted on docker hub for each version. What is the “correct” way to accomplish that using GitHub Actions?</p>

**Comment by Yanjingzhu** [comment]
<p>You could set matrix variable as the value of an environment variable. Then use it in docker build command. Please see my example: </p>
<p><img src="https://user-images.githubusercontent.com/24806478/181088681-4fc98bc0-d935-472e-a458-c47f02c662e5.png" alt="matrix tag.png" data-base62-sha1="7qz44a4l0exZHPwAWdpJ7ubLZUQ" width="657" height="446"></p>
<p>Here is the result: all PHP version tag have been published to register.  </p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/e/e18609cc3d0531ae4a7152ba2230b1cfac3eee98.png" data-download-href="/uploads/short-url/wb4BBu5dAOcyJobfOrnVmqQUlJu.png?dl=1" title="Docker image.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181088682-0b75b2a2-84c1-43a2-9664-930bc5216f0c.png" alt="Docker image.png" data-base62-sha1="wb4BBu5dAOcyJobfOrnVmqQUlJu" width="690" height="317" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/e/e18609cc3d0531ae4a7152ba2230b1cfac3eee98_2_690x317.png, https://aws1.discourse-cdn.com/github/original/2X/e/e18609cc3d0531ae4a7152ba2230b1cfac3eee98.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/e/e18609cc3d0531ae4a7152ba2230b1cfac3eee98.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/e/e18609cc3d0531ae4a7152ba2230b1cfac3eee98_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Docker image.png</span><span class="informations">954×439 31.5 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by lots0logs** [comment]
<p>That works fine when your use case is building a docker image. That’s not my use case though. Based on your proposed work-around, I guess I can start my container manually. The thing is, that is still just a work-around for something that should really be considered a bug in the actions platform. My use case is pretty simple and not an edge case so I hope that your team is working on a proper solution.</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>If you use container in your job, you could use matrix variable as the image name. </p>
<p>Please see my example: </p>
<pre><code>name: matrix
on: push
jobs:
  build:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        version: [5.2, 5.6]
    container: imagename:${{ matrix.version }}
    steps:
    - name: Run a one-line script
      run: echo Hello, world
</code></pre>
<p> If this is still not your case, can you share your scenarios here? </p>

**Reply by mirabilos** [reply]
Doesn’t work for `uses`-style containers though:

```
  "jobs": {
    "eol-dietlibc": {
      "runs-on": "ubuntu-latest",
      "steps": [
        {
          "uses": "actions/checkout@v4.1.1"
        },
        {
          "uses": "docker://debian/eol:${{ matrix.dist }}",
          "with": {
            "args": "sh .extra/deb.sh ${{ matrix.dist }} diet"
          }
        }
      ],
      "strategy": {
        "fail-fast": true,
        "matrix": {
          "dist": [
…
```

**Comment by lots0logs** [comment]
<p>Thanks. That solution is much better. Though it makes things less flexible overall because it requires the containers to have all the dependcies needed to checkout the repos as well as to run any other steps you need. I hope that it will be possible to use matrix variables in the value of “uses” key soon.</p>
